File: PadroesEstruturais/Composite/composite01.py
#Classe Componente
class Componente:
    def __init__(self, nome):
        self.nome = nome
        
    def adicionar(self):
        pass
    
    def remover(self):
        pass
    
    def obterNome(self):
        pass
    

#Classe Folha
class Folha (Componente):
    def __init__(self, nome, preco):
        super().__init__(nome)
        self.preco = preco
        
    def obterNome(self):
        return self.nome
    
    def obterPreco(self):
        return self.preco
    
#Classe Container
class Container(Componente):
    def __init__(self, nome):
        super().__init__(nome)
        self.componentes = []
        
    def adicionar(self, componente):
        self.componentes.append(componente)
        
    def remover(self, componente):
        index = self.componentes.index(componente)
        self.componentes.pop(index)
        
    def obterNome(self):
        return self.nome
    
    def obterPreco(self):
        preco = 0
        for componentes in self.componentes:
            preco += componentes.obterPreco()
        return preco
    
    
#Interface de uso --------------------
maca = Folha('Maçã', 4.50)
laranja = Folha('laranja', 2.90)
uva = Folha('uva', 5.00)    
abacaxi = Folha('Abacaxi', 5.00)

# Criando os # Criando os elementos Embalagem -------
embalagem = Folha('Embalagem 1', 4.00)


# Criando Tabela de Frete --------------
freteSP = Folha('SP', 10.00)
freteRJ = Folha('RJ', 20.00)
freteMG = Folha('MG', 30.00)

# Criando os Containers:
frutas = Container('Frutas')
frutas.adicionar(maca)
frutas.adicionar(laranja)
frutas.adicionar(uva)
frutas.adicionar(abacaxi)


caixa = Container('Caixa de Fruta')
caixa.adicionar(frutas)
caixa.adicionar(embalagem)


frete = Container('Frete - Entrega')
frete.adicionar(caixa)
frete.adicionar(freteRJ)



print(f"Preço Frutas (Apenas as Frutas)....................... R$ {frutas.obterPreco()}")
print(f"Preço Caixa de Fruta (Frutas + Embalagem)..............R$ {caixa.obterPreco()}")
print(f"Preço Frete (Caixa de Fruta + Frete) - Preço Total.....R$ {frete.obterPreco()}")

File: PadroesEstruturais/Composite/composite01.js
//Classe Componente
class Componente {
    constructor(nome) {
        this.nome = nome;
    }

    adicionar(){}

    remover(){}

    obterNome() {}
    
}

//Classe Folha
class Folha extends Componente{
    constructor(nome, preco) {
        super(nome)
        this.preco = preco
    }

    obterNome() {
        return this.nome;
    }

    obterPreco() {
        return this.preco;
    }
}


class Container extends Componente {
    constructor(nome) {
        super(nome);
        this.componentes = [];
    }

    adicionar(componente) {
        this.componentes.push(componente);
    }

    remover(componente) {
        const index = this.componentes.indexOf(componente);
        this.componentes.splice(index, 1);
    }

    obterNome() {
        return this.nome;
    }

    obterPreco() {
        let preco = 0;
        this.componentes.forEach(componente => {
            preco+= componente.obterPreco();
        })

        return preco;
    }
}

// Interface do Usuario - Utilizaçao do Padrão:
const maca = new Folha('Maçã', 4.59);
const laranja = new Folha('Laranja', 2.98);
const uva = new Folha('Uva', 5.00);
const abacaxi = new Folha('Abacaxi', 3.99);

//// A embalagem ao inves de criar como uma variavel, foi inserido direto (abaixo)
const emabalagem = new Folha('Embalagem', 4.00);

//Tabela de Frete
const freteSP = new Folha('SP', 10.00);
const freteRJ = new Folha('RJ', 20.00);
const freteMG = new Folha('MG', 30.00);

const frutas = new Container('Frutas');
frutas.adicionar(maca);
frutas.adicionar(laranja);
frutas.adicionar(uva);

const caixa = new Container('Caixa de Frutas');
caixa.adicionar(frutas);
caixa.adicionar(emabalagem)

//caixa.adicionar(new Folha('Embalagem', 4.00));

const frete = new Container('Frete');
frete.adicionar(caixa);
frete.adicionar(freteMG);


console.log(`Preço Frutas (apenas as frutas): R$ ${frutas.obterPreco()}`)
console.log(`Preço Caixa de Frutas (Frutas + Embalagem): R$ ${caixa.obterPreco()}`)
console.log(`Preço Frete (Caixa de fruta + frete) - Preço Total: R$ ${frete.obterPreco()}`)
